### PR TITLE
Leaderboard

### PR DESCRIPTION
**Pull request by xleonx0x**
### Leaderboard 🏅 🏆 
* Automatically reads database, ranks them, and loads them into the frontend.
* Works for different gamemode types.
* Only reads through games collection so no need for extra submission endpoints (obv will need submit/games endpoint)

### CODE REVIEW

File: backend/src/index.ts [hunk 1]
```diff
@@ -184,6 +192,68 @@ app.post("/login", async (req: TypedRequest<LoginBody>, res: Response) => {
...
+      const id = docSnapshot.id;
+
+      if (!highestScores[username] || score > highestScores[username].score) {
+        highestScores[username] = { id, score };
+      }
+    });
+
+    const ids: string[] = [];
+    for (const usernames in highestScores) {
+      ids.push(highestScores[usernames].id);
```
Comment (lachlanshoesmith) [inline backend/src/index.ts]: `const ids = Object.values(highestScores).map(user => user.id);`

File: frontend/src/_variables.scss [hunk 2]
```diff
@@ -1,6 +1,18 @@
 $dark: hsl(300, 1%, 14%);
+$dark-line: hsl(300, 1%, 35%);
 $navbar-dark: hsl(240, 7%, 14%);
 $light-dark: hsl(0, 0%, 60%);
 $yellow-button: hsl(43, 90%, 40%);
 $primary-yellow: hsl(65, 100%, 60%);
 $secondary-yellow: hsl(52, 100%, 50%);
+$ranking-yellow: hsl(49, 100%, 50%);
+$white-text: #ffffff;
```
Comment (lachlanshoesmith) [inline frontend/src/_variables.scss]: this feels a little redundant given that you can literally specify `white` in css

File: frontend/src/_variables.scss [hunk 3]
```diff
@@ -1,6 +1,18 @@
...
+$exploration: #8eda87;
```
Comment (lachlanshoesmith) [inline frontend/src/_variables.scss]: should also be hsl for consistency

File: frontend/src/components/Forms/Login/Login.tsx [hunk 4]
```diff
@@ -158,11 +158,11 @@ function Login() {
           >
             Register
           </a>
-          <br /> or play as a <a className={classes.link}>guest</a>
+          <br /> or play as a <a className={classes.link} onClick={() => navigate("/gamemodes", {replace: false})}>guest</a>
```
Comment (lachlanshoesmith) [inline frontend/src/components/Forms/Login/Login.tsx]: unnecessary `replace: false`

File: frontend/src/components/Forms/Register/Register.tsx [hunk 5]
```diff
@@ -6,7 +6,7 @@ import WarningText from "../WarningText/WarningText";
 import { useNavigate } from "react-router-dom";
 import { EyeOff, Eye } from "lucide-react";
 
-function Register() {
+function LoginPage() {
```
Comment (lachlanshoesmith) [inline frontend/src/components/Forms/Register/Register.tsx]: should be `RegisterPage`?

File: frontend/src/components/Forms/Register/Register.tsx [hunk 6]
```diff
@@ -232,4 +232,4 @@ function Register() {
   );
 }
 
-export default Register;
+export default LoginPage;
```
Comment (lachlanshoesmith) [inline frontend/src/components/Forms/Register/Register.tsx]: same as above

File: frontend/src/components/Forms/Sheet/Sheet.module.scss [hunk 7]
```diff
@@ -10,7 +10,17 @@
   justify-content: center;
 }
 
-.sheet {
+.containerGamemode {
+  height: 107vh;
+  width: 100vw;
+  opacity: 95%;
+  background: rgba(0, 0, 0, 0.2);
```
Comment (lachlanshoesmith) [inline frontend/src/components/Forms/Sheet/Sheet.module.scss]: using this in conjunction with `opacity` is a bit confusing

File: frontend/src/components/Forms/Sheet/Sheet.tsx [hunk 8]
```diff
@@ -1,15 +1,31 @@
 import classes from "./Sheet.module.scss";
 import { X } from "lucide-react";
+import classNames from "classnames";
 
 interface SheetProps {
   hasCloseButton?: () => void;
   children?: React.ReactNode;
+  sheetLogin?: boolean;
```
Comment (lachlanshoesmith) [inline frontend/src/components/Forms/Sheet/Sheet.tsx]: i feel like these names could be abbreviated to just 'login', 'gamemode', and 'leaderboard'

**Review by lachlanshoesmith — CHANGES_REQUESTED** [review]

File: frontend/src/components/Forms/Sheet/Sheet.module.scss (hunk 7 above)
Comment (xleonx0x) [reply frontend/src/components/Forms/Sheet/Sheet.module.scss]: background just darkens the pano a bit

**Comment by xleonx0x** [comment]
i would merge/review gamemodes first

**Review by lachlanshoesmith — DISMISSED** [review]

**Comment by xleonx0x** [comment]
ill resolve the conflicts

**Comment by lachlanshoesmith** [comment]
ill fix conflicts

**Comment by lachlanshoesmith** [comment]
> ill resolve the conflicts

ok nvm go ahead

**Comment by xleonx0x** [comment]
@lachlanshoesmith the title looks kinda misaligned with the other navbar items can you take a look 🙏

**Review by lachlanshoesmith — APPROVED** [review]

**Pull request merged**